contracts/utils/IdentityGuard.sol

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { RONTransferHelper } from "../extensions/RONTransferHelper.sol";
+import { AddressArrayUtils } from "../libraries/AddressArrayUtils.sol";
+import { IERC165 } from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import { ErrZeroAddress, ErrOnlySelfCall, ErrZeroCodeContract, ErrNonpayableAddress, ErrUnsupportedInterface } from "./CommonErrors.sol";
+
+abstract contract IdentityGuard is RONTransferHelper {
+  using AddressArrayUtils for address[];
+
+  /**
+   * @dev Modifier to restrict functions to only be called by this contract.
+   * @dev Reverts if the caller is not this contract.
+   */
+  modifier onlySelfCall() virtual {
+    _requireSelfCall();
+    _;
+  }
+
+  /**
+   * @dev Modifier to ensure that the elements in the `arr` array are non-duplicates.
+   * It calls the internal `_checkDuplicate` function to perform the duplicate check.
+   *
+   * Requirements:
+   * - The elements in the `arr` array must not contain any duplicates.
+   */
+  modifier nonDuplicate(address[] memory arr) virtual {
+    _requireNonDuplicate(arr);
+    _;
+  }
+
+  /**
+   * @dev Internal method to check the method caller.
+   * @dev Reverts if the method caller is not this contract.
+   */
+  function _requireSelfCall() internal view virtual {
+    if (msg.sender != address(this)) revert ErrOnlySelfCall(msg.sig);
+  }
+
+  /**
+   * @dev Internal function to check if a contract address has code.
+   * @param addr The address of the contract to check.
+   * @dev Throws an error if the contract address has no code.
+   */
+  function _requireHasCode(address addr) internal view {
+    if (addr.code.length == 0) revert ErrZeroCodeContract(addr);
+  }
+
+  /**
+   * @dev Checks if an address non-payable and reverts if it is.
+   * @param addr The address to check.
+   */
+  function _requirePayableAddress(address addr) internal {
+    if (!_unsafeSendRON(payable(addr), 0, 0)) {
+      revert ErrNonpayableAddress(addr);
+    }
```

Comment:
Please write some unit test to ensure it works as expected. Looks like this only works with EOA account, but the contract account with receive method will be reverted.

Reply:
Correct me if I'm wrong. This instruction triggers [opCall](https://github.com/axieinfinity/ronin/blob/c0d708a7806d1eaec4ae98922bf8eb35375ac96d/core/vm/instructions.go#L648) and jumps into [EVM CALL](https://github.com/axieinfinity/ronin/blob/c0d708a7806d1eaec4ae98922bf8eb35375ac96d/core/vm/evm.go#L208): https://github.com/axieinfinity/ronin/blob/c0d708a7806d1eaec4ae98922bf8eb35375ac96d/core/vm/evm.go#L275-L290

- The EOA accounts can bypass because the gas doesn't change
- The contract accounts will be reverted with the error code "out of gas" because the EVM triggers the receive method in the contract and [costs an operation gas](https://github.com/axieinfinity/ronin/blob/c0d708a7806d1eaec4ae98922bf8eb35375ac96d/core/vm/interpreter.go#L226-L228)

Reply:
Please note that because the call value is 0, the gas won't have stipend gas (2300): https://github.com/axieinfinity/ronin/blob/c0d708a7806d1eaec4ae98922bf8eb35375ac96d/core/vm/instructions.go#L666-L669

Reply:
`IdentityGuard:_requirePayableAddress` is removed since the guard is uneffective.